File: api/app/routers/feeder.py
import os
import httpx
from fastapi import APIRouter, Depends, UploadFile, File, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from app.db.database import get_db
from app.services.feeder_actions import handle_feed_cat_default
from app.clients.feeder import get_stream_url
from app.schemas.feeder import FeederEvent
from app.utils.feeder_messages import EVENT_MESSAGES
from datetime import datetime, timedelta

router = APIRouter(prefix="/feeder", tags=["feeder"])

_last_api_error: datetime | None = None

DISCORD_TOKEN = os.getenv("DISCORD_TOKEN")
DISCORD_ALERT_CHANNEL_ID = int(os.getenv("DISCORD_ALERT_CHANNEL_ID", "0"))


async def send_discord_message(
    content: str, file_bytes: bytes | None = None, filename: str = "snapshot.jpg"
):
    """Push a message (optionally with image) to Discord channel via REST."""
    headers = {"Authorization": f"Bot {DISCORD_TOKEN}"}
    url = f"https://discord.com/api/v10/channels/{DISCORD_ALERT_CHANNEL_ID}/messages"

    async with httpx.AsyncClient() as client:
        if file_bytes:
            files = {
                "file": (filename, file_bytes, "image/jpeg"),
                "payload_json": (None, f'{{"content": "{content}"}}'),
            }
            await client.post(url, headers=headers, files=files)
        else:
            await client.post(url, headers=headers, json={"content": content})


@router.post("/motion")
async def motion_detected(snapshot: UploadFile = File(...)):
    """
    Called by ESP32 when motion sensor triggers.
    Receives a JPEG snapshot and pushes alert + stream URL to Discord.
    """
    image_bytes = await snapshot.read()
    stream_url = get_stream_url()

    message = f"🐱 Wykryto ruch przy karmiku!\n📹 Podgląd na żywo: {stream_url}"

    try:
        await send_discord_message(content=message, file_bytes=image_bytes)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Discord push failed: {e}")

    return {"status": "ok"}


@router.post("/feed/default")
async def feed_default(db: AsyncSession = Depends(get_db)):
    """
    Trigger feeding with default config.
    Called by: NFC tag, Google Assistant webhook.
    No AI involved — straight to ESP32.
    """
    result = await handle_feed_cat_default(db)
    return {"status": "ok", "message": result}


@router.post("/event")
async def feeder_event(event: FeederEvent):
    global _last_api_error

    if event.type == "api_error":
        now = datetime.now()
        if _last_api_error and now - _last_api_error < timedelta(minutes=60 * 24):
            return {"status": "ok"}
        _last_api_error = now

    message = EVENT_MESSAGES.get(event.type, f"ℹ️ Karmnik: {event.type}")
    try:
        await send_discord_message(content=message)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Discord push failed: {e}")
    return {"status": "ok"}


@router.get("/nfc")
async def feed_nfc(token: str, db: AsyncSession = Depends(get_db)):
    """GET endpoint for NFC tag — triggers default feeding."""
    if token != os.getenv("FEEDER_PASS_NFC"):
        raise HTTPException(status_code=403, detail="Forbidden")
    await handle_feed_cat_default(db)
    return {"status": "ok"}

File: api/src/routers/feeder.ts
import { Router, Request, Response } from "express";
import multer from "multer";
import axios from "axios";
import FormData from "form-data";
import { handleFeedCatDefault } from "../services/feederActions";
import { getStreamUrl } from "../clients/feeder";
import { EVENT_MESSAGES } from "../utils/feederMessages";
import { MESSAGES } from "../messages";

const router = Router();
const upload = multer({ storage: multer.memoryStorage() });

const DISCORD_TOKEN = process.env["DISCORD_TOKEN"];
const DISCORD_ALERT_CHANNEL_ID = process.env["DISCORD_ALERT_CHANNEL_ID"] ?? "0";

let lastApiError: Date | null = null;

async function sendDiscordMessage(content: string, fileBytes?: Buffer, filename = "snapshot.jpg") {
  const url = `https://discord.com/api/v10/channels/${DISCORD_ALERT_CHANNEL_ID}/messages`;
  const headers = { Authorization: `Bot ${DISCORD_TOKEN}` };

  if (fileBytes) {
    const form = new FormData();
    form.append("file", fileBytes, { filename, contentType: "image/jpeg" });
    form.append("payload_json", JSON.stringify({ content }));
    await axios.post(url, form, { headers: { ...headers, ...form.getHeaders() } });
  } else {
    await axios.post(url, { content }, { headers });
  }
}

router.post("/motion", upload.single("snapshot"), async (req: Request, res: Response) => {
  if (!req.file) return res.status(400).json({ detail: MESSAGES.NO_SNAPSHOT });

  const streamUrl = getStreamUrl();
  try {
    await sendDiscordMessage(MESSAGES.FEEDER_MOTION_ALERT(streamUrl), req.file.buffer);
    return res.json({ status: "ok" });
  } catch (e) {
    return res.status(500).json({ detail: MESSAGES.DISCORD_PUSH_FAILED(e) });
  }
});

router.post("/feed/default", async (_req: Request, res: Response) => {
  const message = await handleFeedCatDefault();
  return res.json({ status: "ok", message });
});

router.post("/event", async (req: Request, res: Response) => {
  const { type } = req.body as { type: string };

  if (type === "api_error") {
    const now = new Date();
    if (lastApiError && now.getTime() - lastApiError.getTime() < 1000 * 60 * 60 * 24) {
      return res.json({ status: "ok" });
    }
    lastApiError = now;
  }

  const message = EVENT_MESSAGES[type] ?? MESSAGES.FEEDER_EVENT_UNKNOWN(type);
  try {
    await sendDiscordMessage(message);
    return res.json({ status: "ok" });
  } catch (e) {
    return res.status(500).json({ detail: MESSAGES.DISCORD_PUSH_FAILED(e) });
  }
});

router.get("/nfc", async (req: Request, res: Response) => {
  if (req.query["token"] !== process.env["FEEDER_PASS_NFC"]) {
    return res.status(403).json({ detail: MESSAGES.FORBIDDEN });
  }
  await handleFeedCatDefault();
  return res.json({ status: "ok" });
});

export default router;
